cli/pcluster/config/resource_map.py

```diff
@@ -0,0 +1,128 @@
+            """
+            # First purge removed values from internal resources
+            self.__resources = [v if v in values else None for v in self.__resources]
+
+            # Then add only new values
+            for value in list(filter(lambda l: l not in self.__resources, values)):
+                # Assign first free slot to the label
+                for i in range(len(self.__resources)):
+                    if not self.__resources[i]:
+                        self.__resources[i] = value
+                        break
+                else:
+                    raise Exception("No more resources available for value {0}".format(value))
+
+        def resources(self):
+            """Return the resources values in this array."""
+            return self.__resources
+
+    def __init__(self, initial_data=None):
+        """
+        Create a ResourceMap.
+
+        A ResourceMap can be constructed from scratch (like in the case of a new configuration file) or starting from a
+        dictionary representing the resources structure, like the one that can be obtained by calling the resources()
+        method.
+
+        :param initial_data: A dict containing the initial data for the structure, or None if a new empty ResourceMap is
+        being created
+        """
+        self.__resource_arrays = {}
+        if initial_data:
+            for key, resources in initial_data.items():
+                self.__resource_arrays[key] = ResourceMap.ResourceArray(resources)
+
+    def resources(self, key=None):
+        """
+        Return the resources array corresponding to the provided key.
+
+        If no key is provided, it returns a dict containing all the Resource arrays. The purpose of this method is to
+        return a serializable structure that can be stored somewhere (ex. in a CloudFormation template) and deserialized
+        back into a ResourceMap by passing it to ResourceMap's constructor.
+
+        :param key: The key of the resources array to be returned, or None if the all structure is needed.
+        :return: The requested resources array or a dict with all the resources arrays
+        """
+        if not key:
+            resources_map = {}
+            for key, resource_array in self.__resource_arrays.items():
+                resources_map[key] = resource_array.resources()
+            return resources_map
+        else:
+            resource_array = self.__resource_arrays.get(key)
+            return resource_array.resources() if resource_array else None
+
+    def store(self, key, values):
+        """
+        Store the provided values in the resources array corresponding to the specified key.
+
+        Values will be stored in the resources array according to the rules specified in ResourceArray.store() method.
+
+        :param key: The key of the resources array
+        :param values: The values to store.
+        """
+        self.__resource_arrays.get(key).store(values)
```

Comment:
The spirit of the class is that you must always explicitly allocate the resource before storing the values in it. This is because we need to know in advance how many resource "slots" are needed.